lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/NetworkSpec.hs

```diff
@@ -0,0 +1,384 @@
+    -- chaff is the same shape as the node chain, but with different hashes
+    chaff = [bh { prevBlockHash = Hash "x" } | bh <- nodeChain]
+
+-- | The slot index at which the local chain starts.
+firstSlot :: [BlockHeader] -> Maybe SlotId
+firstSlot bs
+    | hasTip bs = slotId <$> headMay bs
+    | otherwise = Nothing
+
+dropToSlot :: SlotId -> [BlockHeader] -> [BlockHeader]
+dropToSlot sl = dropWhile ((< sl) . slotId)
+
+takeToSlot :: SlotId -> [BlockHeader] -> [BlockHeader]
+takeToSlot sl = takeWhile ((< sl) . slotId)
+
+{-------------------------------------------------------------------------------
+                              Test data generation
+-------------------------------------------------------------------------------}
+
+-- | Generate an infinite test chain. Take a slice of the list and use 'tipId'
+-- and 'headerIds' to access the tip and block headers. The tip of a test chain
+-- is the penultimate block.
+chain :: String -> [BlockHeader]
+chain p = [BlockHeader (SlotId 0 n) (Hash . B8.pack $ p ++ hash n) | n <- [0..]]
+  where
+    hash 0 = ""
+    hash n = show (n - 1)
+
+-- | Filter out test chain blocks that correspond to False values, and update
+-- parent hashes so that the chain is still continuous.
+removeBlocks :: [Bool] -> [BlockHeader] -> [BlockHeader]
+removeBlocks holes bs =
+    reverse $ snd $
+    foldl' filt (prevBlockHash (head bs), []) $
+    zip holes (zip (map prevBlockHash $ tail bs) bs)
+  where
+    -- make holes and update prev header hashes
+    filt (prev, ac) (True, (h, BlockHeader sl _))
+        = (h, ((BlockHeader sl prev):ac))
+    filt pbs _ = pbs
+
+instance Arbitrary TestCase where
+    arbitrary = do
+        k <- arbitrary
+        let kk = fromIntegral k' where Quantity k' = k
+        -- Choose an arbitrary local length proportional to k
+        localChainLength <- choose (0, kk * 4)
+        -- Choose where the local unstable block headers start from
+        localChainStart <- (localChainLength -) <$> choose (0, kk)
+        -- Choose an arbitrary node chain length proportional to k
+        nodeChainLength <- choose (0, kk * 4)
+        -- Generate some empty slots - False means empty
+        let genHole = frequency [(1, pure False), (4, pure True)]
+            genHoles n = (True:) <$> vectorOf (n - 1) genHole
+        emptyA <- genHoles localChainLength
+        emptyB <- genHoles nodeChainLength
+        -- At some point the node chain changes from a to b
+        forkPoint <- SlotId 0 . fromIntegral <$>
+            choose (localChainStart, localChainLength - 1)
+        let chainA = removeBlocks emptyA $ chain "a"
+            chainB = removeBlocks emptyB $ chain "b"
+            chainAB = takeToSlot forkPoint chainA ++ dropToSlot forkPoint chainB
+        chainB' <- elements [ chainA, chainAB ]
+
+        let localChain = take localChainLength $ drop localChainStart chainA
+            nodeChain = take nodeChainLength chainB'
```

Comment:
I'd suggest splitting the above into some intermediate `Arbitrary` instances working with a `newtype Chain (k :: Nat) = Chain [Block]` or similar and leave the `TestCase` to be a composition of the smaller pieces. 
If I understand correctly, to build the above we need a chain generator that can construct chains with "holes" (i.e. empty slots). So this can be done by default in the chain generator. 

Then, we can create a test case by:

```hs

-- Creating an arbitrary chain `base`
base <- arbitrary :: Arbitrary (Chain k) 

-- Creating a local extension `local`
localExt <- arbitrary :: Arbitrary (Chain k) 

-- Creating a remote extension `remote`
remoteExt <- arbitrary :: Arbitrary (Chain k) 

-- Compose them together to create two chains with a fork point:
let localChain = base <> localExt
let remoteChain = base <> remoteExt
```

with `remoteExt` and `localExt` possibly empty? What do you think?